File: 2term.labs/lab 2.2/c++/Trapezoid.cpp
#include "Trapezoid.h"
#include <iostream>
#include <cmath>

using namespace std;

double distance(Vertex A, Vertex B);

Trapezoid::Trapezoid(Vertex _A, Vertex _B, Vertex _C, Vertex _D) : A(_A), B(_B), C(_C), D(_D) {};

Vertex Trapezoid::getA() const {
    return A;
}

Vertex Trapezoid::getB() const {
    return B;
}

Vertex Trapezoid::getC() const {
    return C;
}

Vertex Trapezoid::getD() const {
    return D;
}

double Trapezoid::getArea() const {
    double l1 = distance(A, D);
    double l2 = distance(B, C);

    double h = fabs((B.y - A.y) * A.x - (B.x - A.x) * A.y + (B.x * A.y - A.x * B.y)) / l1;

    double area = ((l1 + l2) * h) / 2;
    return area;
}

double Trapezoid::getPerimeter() const {
    double l1 = distance(A, B);
    double l2 = distance(B, C);
    double l3 = distance(C, D);
    double l4 = distance(D, A);
    double perimeter = l1 + l2 + l3 + l4;
    return perimeter;
}

double distance(Vertex A, Vertex B) {
    double distance = sqrt(pow(B.x - A.x, 2) + pow(B.y - A.y, 2));
    return distance;
};

File: 2term.labs/lab 2.2/c#/Trapezoid.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ClassLibrary
{
    public struct Vertex
    {
        public int x, y;

        public Vertex(int x, int y)
        {
            this.x = x;
            this.y = y;
        }
    }
    public class Trapezoid : Shape
    {
        public Vertex A { get; private set; }
        public Vertex B { get; private set; }
        public Vertex C { get; private set; }
        public Vertex D { get; private set; }

        public Trapezoid(Vertex A, Vertex B, Vertex C, Vertex D)
        {
            this.A = A;
            this.B = B;
            this.C = C;
            this.D = D;
        }

        public double Distance(Vertex A, Vertex B)
        {
            double distance = Math.Sqrt(Math.Pow(B.x - A.x, 2) + Math.Pow(B.y - A.y, 2));
            return distance;
        }

        public override double GetArea()
        {
            double a = D.y - A.y;
            double b = A.x - D.x;
            double c = D.x * A.y - A.x * D.y;

            double numerator = Math.Abs(a * B.x + b * B.y + c);
            double denominator = Math.Sqrt(a * a + b * b);
            double h = numerator / denominator;
            double l1 = Distance(A, D);
            double l2 = Distance(B, C);

            double area = ((l1 + l2) * h) / 2;
            return area;
        }

        public override double GetPerimeter()
        {
            double l1 = Distance(A, B);
            double l2 = Distance(B, C);
            double l3 = Distance(C, D);
            double l4 = Distance(D, A);
            double perimeter = l1 + l2 + l3 + l4;
            return perimeter;
        }

    }
}
